File: lib/functions/generate.dart
import 'dart:io';

import 'package:bliss/compiler/categorizer.dart';
import 'package:bliss/compiler/emitter.dart';
import 'package:bliss/compiler/tokenizer.dart';
import 'package:bliss/constants.dart';
import 'package:bliss/functions/stdlib.dart';
import 'package:cli_spin/cli_spin.dart';

(String, String) generate(String inputFileName, {bool noStdlib = false}) {
  final inputFile = File(inputFileName);
  final contents = inputFile.readAsStringSync();

  final lastDotIdx = inputFileName.lastIndexOf(".");
  final inputFileNameWithoutExt = inputFileName.substring(0, lastDotIdx);
  final outputFileNameWithoutExt = "$buildDir/$inputFileNameWithoutExt";

  final irOutputFileName = "$outputFileNameWithoutExt.re";
  final exeOutputFileName = "$outputFileNameWithoutExt.exe";

  print("==> Generating: $inputFileName -> $irOutputFileName");

  final spinner = CliSpin(spinner: CliSpinners.dots);
  spinner.start();

  spinner.text = "Tokenizing source...";

  final tokens = tokenize(contents);

  spinner.text = "Parsing tokens...";

  final stdlibAst = noStdlib ? null : generateStdlibAst();
  final ast = Categorizer().categorize(tokens);

  spinner.text = "Emitting IR...";

  final ir = Emitter().emit(ast, stdlibAst: stdlibAst);
  final irWithStdlib = ir;

  File(irOutputFileName)
    ..createSync(recursive: true)
    ..writeAsStringSync(irWithStdlib);

  spinner.success("Generated IR: $colorCyan$irOutputFileName$colorReset");

  return (irOutputFileName, exeOutputFileName);
}

String generateDry(String input, {bool noStdlib = false}) {
  final tokens = tokenize(input);

  final stdlibAst = noStdlib ? null : generateStdlibAst();
  final ast = Categorizer().categorize(tokens);

  final ir = Emitter().emit(ast, stdlibAst: stdlibAst, noFormatting: true);

  return ir;
}

File: src/Functions/Generate.hs
module Functions.Generate (generate) where

import Compiler.Categorizer (categorize)
import Compiler.Emitter (emit)
import Compiler.Tokenizer (tokenize)
import Constants (buildDir)
import System.Directory (createDirectoryIfMissing, exeExtension)
import Utils (lastIndexOf)

generate :: String -> IO (String, String)
generate inputFileName = do
  contents <- readFile inputFileName

  let lastDotIdx = lastIndexOf '.' inputFileName
  let (inputFileNameWithoutExt, _) = splitAt lastDotIdx inputFileName
  let outputFileNameWithoutExt = buildDir ++ "/" ++ inputFileNameWithoutExt

  let lastSlashIdx = lastIndexOf '/' outputFileNameWithoutExt
  let (outputDir, _) = splitAt lastSlashIdx outputFileNameWithoutExt

  let irOutputFileName = outputFileNameWithoutExt ++ ".re"
  let exeOutputFileName = outputFileNameWithoutExt ++ exeExtension

  putStrLn $ "==> Generating: " ++ inputFileName ++ " -> " ++ irOutputFileName

  putStrLn "   🔹 Tokenizing source"

  let tokens = tokenize contents

  putStrLn "   🔹 Parsing tokens"

  let ast = categorize tokens

  putStrLn "   🔹 Emitting IR"

  let ir = emit ast

  createDirectoryIfMissing True outputDir
  writeFile irOutputFileName ir

  putStrLn $ "   ✅️ Generated IR: " ++ irOutputFileName

  -- putStrLn "TOKENS:"
  -- print tokens
  -- putStrLn "AST:"
  -- print ast
  -- putStrLn "IR:"
  -- putStrLn ir

  return (irOutputFileName, exeOutputFileName)
